WebXR Plane Detection - User Manual (#613)
playcanvas/developer.playcanvas.com: docs/user-manual/xr/ar/plane-detection.md

---
title: Plane Detection
sidebar_position: 1
---

Very similar to [Mesh Detection][2], Plane Detection provides access to planes that estimate real-world surfaces.

Each plane can represent a surface with optionally available labels, such as: walls, floor, table, etc.

The underlying system can provide pre-captured data as well as provide real-time reconstruction depending on the underlying system implementation.

## Support

```javascript
if (app.xr.planeDetection.supported) {
    // plane detection is supported
}

app.xr.on('start', () => {
    if (app.xr.planeDetection.available) {
        // plane detection is available
    }
});
```

## Access

A feature flag needs to be added to the session start:

```javascript
app.xr.start(camera, pc.XRTYPE_AR, pc.XRSPACE_LOCALFLOOR, {
    planeDetection: true
});
```

Then planes are added/removed in async:

```javascript
app.xr.planeDetection.on('add', (xrPlane) => {
    // a new XrPlane has been added

    xrPlane.once('remove', () => {
        // an XrPlane has been removed
    });
});
```

Also, the list of XrPlane'es is available:

```javascript
const xrPlanes = app.xr.planeDetection.meshes;

> for(let i = 0; i < xrPlanes.length; i++) {

Reviewer: ```suggestion
for (let i = 0; i < xrPlanes.length; i++) {
```

    const xrPlane = xrPlanes[i];
}
```

## Mesh

> Each plane is represented as an instance of XrPlane, which can be added/removed during an active session. It also has data that can be changed during its lifetime.

Reviewer: ```suggestion
Each plane is represented as an instance of XrPlane which can be added/removed during an active session. It also has data that can be changed during its lifetime.
```

> You can access the position and rotation (world space) of an XrPlane:

Reviewer: ```suggestion
You can access the position and rotation (world-space) of an XrPlane:
```

```javascript
entity.setPosition(xrPlane.getPosition());
entity.setRotation(xrPlane.getRotation());
```

> Each XrPlane has its points (in local space) and orientation, that can be used to construct a visual mesh. The list of points represents the vertices of a polygon perimeter.

Reviewer: ```suggestion
Each XrPlane has its points (in local-space) and orientation, that can be used to construct a visual mesh. The list of points represents the vertices of a polygon perimeter.
```

> An `xrPlane.orientation` provides information if a plane is vertical, horizontal or anything else.

Reviewer: ```suggestion
An `xrPlane.orientation` provides information as to whether a plane is vertical, horizontal or anything else.
```

    for (let i = 0; i < xrPlane.points.length; i++) {
        vertices[i * 3 + 3 + 0] = xrPlane.points[i].x;
        vertices[i * 3 + 3 + 1] = xrPlane.points[i].y;
        vertices[i * 3 + 3 + 2] = xrPlane.points[i].z;
        indices[i * 3 + 0] = 0;
        indices[i * 3 + 1] = i + 1;
        indices[i * 3 + 2] = ((i + 1) % xrPlane.points.length) + 1;
    }

    mesh.setPositions(vertices); // set vertices
    mesh.setNormals(pc.calculateNormals(vertices, indices)); // calculate normals
    mesh.setIndices(indices); // set indices
    mesh.update(pc.PRIMITIVE_TRIANGLES); // update buffers

    const material = new pc.StandardMaterial();
    const meshInstance = new pc.MeshInstance(mesh, material);

    const entity = new pc.Entity(xrPlane.label);

    // add render component
    entity.addComponent('render', {
        meshInstances: [ meshInstance ]
    });

    // add entity to the scene root
    app.root.addChild(entity);

    // clean up after XrPlane is removed
    xrPlane.once('remove', () => {
        material.destroy();
        mesh.destroy();
        entity.destroy();
    });
});
```

## Semantic Label

> XrPlane can represent various real-world objects, and a label can help to identify what it represents using its property `XrPlane.label`.

Reviewer: ```suggestion
An XrPlane can represent various real-world objects and a label can help to identify what it represents using its property `XrPlane.label`.
```

> These labels can be any of: floor, wall, door, window, table, screen, global mesh, other, and more. Here is a [list of semantic labels][1], although this list is not definitive and the platform can report anything it feels fit.

Reviewer: ```suggestion
These labels can be any of: `floor`, `wall`, `door`, `window`, `table`, `screen`, `global mesh`, `other` and more. Here is a [list of semantic labels][2], although this list is not definitive and the platform can report anything it feels fit.
```

## Changes

> Depending on the underlying system capabilities, the XrPlane attributes can change while an XR session is active. You can subscribe to that event and update a visual mesh accordingly:

Reviewer: ```suggestion
Depending on the underlying system capabilities, XrPlane attributes can change while an XR session is active. You can subscribe to that event and update a visual mesh accordingly:
```

```javascript
xrPlane.on('change', () => {

>     // points and/or label has been changed

Reviewer: ```suggestion
    // points and/or label have been changed
```

});
```

> [1]: https://github.com/immersive-web/semantic-labels/blob/master/labels.json
> [2]: /user-manual/xr/ar/mesh-detection/

Reviewer: ```suggestion
[1]: /user-manual/xr/ar/mesh-detection/
[2]: https://github.com/immersive-web/semantic-labels/blob/master/labels.json
```
